add textarea, and convert note

mallezie opened the pull request:
Add textarea support to addfield, and convert note / relation form. 
Remove unused field from that form 
Small update to https://issues.civicrm.org/jira/browse/CRM-16178

colemanw commented on CRM/Contact/Form/Edit/Notes.php:
```diff
@@ -42,8 +42,7 @@ class CRM_Contact_Form_Edit_Notes {
    */
   public static function buildQuickForm(&$form) {
     $form->applyFilter('__ALL__', 'trim');
-    $form->add('text', 'subject', ts('Subject'), array('size' => 60, 'maxlength' => 254));
```
are you sure this field is unused? Notes appear to have a subject.
![newnote](https://cloud.githubusercontent.com/assets/2874912/6885410/919c8e08-d5ed-11e4-8b62-9e67c88c50cd.png)

mallezie commented:
True, note's have asubject. But this forms is for the note's on an relationship. (Add relationship, the note-part there, and there there is no relationship. 
The method i'm changing here CRM_Contact_Form_Edit_Notes::buildQuickForm($this);
is only used on the relationship form. 
![note-on-relation](https://cloud.githubusercontent.com/assets/1926222/6885493/9c516c76-d622-11e4-8dc5-8cf3c6356ae3.png)
 
The form you're showing is actually CRM_Note_Form_Note (which is untouched). 
I'm not sure however what's happening on the failed test, i'll need to figure that further out.

mallezie commented:
hmm, as far as i see is this a completely unrelated test-failure, not sure what caused this.

colemanw commented:
yes that test has been sporadically failing and we haven't yet figured out why. It's irritating but harmless.

mallezie commented:
The subject seems to be needed, you were right. It's used in the notes-tab when you add a new contact. My IDE didn't find it since it didn't recognize the usage their. 
I also added the subject html type to the schema.

colemanw commented on CRM/Contact/Form/Edit/Notes.php:
```diff
@@ -42,8 +42,8 @@ class CRM_Contact_Form_Edit_Notes {
-    $form->add('text', 'subject', ts('Subject'), array('size' => 60, 'maxlength' => 254));
-    $form->add('textarea', 'note', ts('Notes'), array('cols' => '60', 'rows' => '3'));
+    $form->addField('subject', array('entity' => 'note', 'size' => 60, 'maxlength' => 254));
+    $form->addField('note', array('entity' => 'note', 'cols' => 60, 'rows' => 3));
```
Rows and cols are already in the schema for this field.

colemanw commented on CRM/Contact/Form/Edit/Notes.php:
```diff
@@ -42,8 +42,8 @@ class CRM_Contact_Form_Edit_Notes {
     $form->applyFilter('__ALL__', 'trim');
-    $form->add('text', 'subject', ts('Subject'), array('size' => 60, 'maxlength' => 254));
-    $form->add('textarea', 'note', ts('Notes'), array('cols' => '60', 'rows' => '3'));
+    $form->addField('subject', array('entity' => 'note', 'size' => 60, 'maxlength' => 254));
```
Can you try leaving out the size & maxlength attributes? I'm pretty sure these are in the schema.

colemanw commented:
FYI you can go to the api explorer and use getfields (or getfield) to see what field attributes are already defined. If a field is missing any attributes in the schema (especially the `html` attribute) I think adding them should be a priority.
(remember to rerun GenCode.php to see your changes take effect)
I see checking/cleaning our field metadata as the biggest benefit of this addField project.

mallezie commented:
I removed the cols from the schema. That worked. I left the rows, since the schema defines the default of 4 rows, and here we use 3, i didn't want to change that. 
Same for the subject. The schema defines a size of 45, and maxlenght (equal to length of 255), here we had resp 60 and 254, so i left that as is. 
while testing i noted i can't change the size and maxlength in the schema tho'se seem overwritten by defaults. (Generated maxlength comes from length, and size get's defined as ``` 'size' => CRM_Utils_Type::HUGE,```)

colemanw commented:
I think when it's "close enough" it would be good to fall back on what's in the schema.
I would remove the hard-coded size of 254 since there's probably no good reason for it.

Size can be defined in the schema like so: https://github.com/colemanw/civicrm-core/blob/4.6/xml/schema/Contact/Contact.xml#L191

If length can't be specified within the `<html>` property in the schema, then that's a bug.

mallezie commented:
I converted the maxsize to the schema value. I also added support for size. 
Is still had to add the size value in addField, although it's added to the schema. The api-call ($fieldspec value) always returns 45 here instead of the defined value of 60. Not sure where that comes from, but that should  be another issue?

mallezie commented:
I rebased this PR

colemanw commented on CRM/Contact/Form/Edit/Notes.php:
```diff
@@ -42,8 +42,8 @@ class CRM_Contact_Form_Edit_Notes {
-    $form->add('text', 'subject', ts('Subject'), array('size' => 60, 'maxlength' => 254));
-    $form->add('textarea', 'note', ts('Notes'), array('cols' => '60', 'rows' => '3'));
+    $form->addField('subject', array('entity' => 'note', 'size' => '60'));
+    $form->addField('note', array('entity' => 'note', 'rows' => 3));
```
Ok sorry to nitpick but why are 'size' and 'rows' not coming from the schema here?

mallezie replied:
No worry, those nitpciks are always valuable reviews. 
For the rows is the default value (which is inside the schema) 4 cols. This is for most textarea's the case, but here it is (and was) 3 cols, so it's overriden there. 
The size of the subject is defined at 60 in the schema, however the getfields api-call returns 45.
This seems because of the schema conversion to the DAO class. The size is defined as 60 in the schema, but gets converted to ```'size' => CRM_Utils_Type::HUGE```, in the generated class, and that constants value is 45. 
The constant class has a todo ```// @todo What's the point of these constants? Backwards compatibility?``` and i think this whole class needs some form of refactoring, so i would leave that here as is, and do that for a follow-up

mallezie commented:
Anything left to do here?

colemanw commented:
Looks great, thanks.
